telemetry_package/main.py

```diff
@@ -1,31 +1,140 @@
-import sys
+'''
+This code gets too many messages all at once. Hence it needs to: 
+    - The Node only sends the first message it receives
+'''
+
+
+import json
+import os
+import select
+import socket
+# import sys
 import rclpy
 from rclpy.node import Node
+from rcl_interfaces.msg import Log
+
+
+DEFAULT_UNIX_SOCKET_PATH = "/tmp/thi_drone"
+
+
+class MinimalPublisher(Node):
 
-from common_package_py.common_node import CommonNode
+    def __init__(self):
+        super().__init__('telemetry_node')
+        self.subscription = self.create_subscription(
+            Log,
+            '/rosout',
+            self.run_server, 
+            100)
+        self.subscription 
 
+        # Counter for debugging purposes
+        self.msg_cnt = 0
 
-class TelemetryNode(CommonNode):
-    """A class that represents a telemetry node.
-    """
+        self.msg_cache = list()
+        self.cache_size = 400
 
-    def __init__(self, id: str):
-        """Creates a new TelemetryNode.
 
-        Args:
-            id (str): Unique node id
+    def is_socket_closed(self, sock: socket.socket) -> bool:
+        try:
+            # this will try to read bytes without blocking and also without removing them from buffer (peek only)
+            data = sock.recv(16, socket.MSG_DONTWAIT | socket.MSG_PEEK)
+            if len(data) == 0:
+                return True
+        except BlockingIOError:
+            return False  # socket is open and reading from it would block
+        except ConnectionResetError:
+            return True  # socket was closed for some other reason
+        except Exception as e:
+            self.get_logger().error(f"Unexpected exception when checking if the socket is closed: {e}")
+            return False
+        return False
+    
+
+    '''
+    Code to implement the websocket functionality     
+    '''
+    def socket_exists(self, socket_path: str) -> bool:
+        return os.path.exists(socket_path)
+
+
+    def send_message(self, client_socket):
         """
+        Defines the message block to be continuosly send.
+        """
+        if len(self.msg_cache) > 0:
+            client_socket.sendall(json.dumps(self.msg_cache[0]).encode())
+            self.get_logger().info(f'Send message #{self.msg_cnt + 1}')
+            self.msg_cnt += 1     
+            self.msg_cache.pop(0)  
+
+
+    def run_server(self, msg, socket_path=DEFAULT_UNIX_SOCKET_PATH):
+        # Check if socket is in use
+        if self.socket_exists(socket_path):
+            self.get_logger().fatal(f'Socket path already exists')
+            raise FileExistsError('')
+
+        # Create a socket
+        server_socket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
```

Comment:
You should create the socket once in your constructor and remove the socket in your destructor.
While being active you constantly keep your socket open.
Opening sockets is relatively expensive and that might bring you into trouble.

Reply:
Now I'm getting what you're doing here: You're opening a **websocket** here.

This changes a few things, but nevertheless you should keep the websocket open as long as possible.

Just check once in your callback if the websocket is still alive and if not, try to open a new one. Otherwise you leave it running.

Reply:
I still have not committed a newer version of the code since I'm still working on it. The socket will be set up before the ROS2 node starts "spinning", so that the socket gets opened only once. This approach resolves many issues including broken pipe errors and the program trying to reopen the socket while it is still in use (and thus ends up crashing).